standard/classes.md

```diff
@@ -5584,3 +5466,146 @@ When the body of the async function terminates, the return task is moved out of
+- When an exception is thrown and propagated out of the iterator block:
+  - Appropriate `finally` blocks in the iterator body will have been executed by the exception propagation.
+  - The state of the enumerator object is changed to **after**.
+  - The exception propagation continues to the caller of the `MoveNext` method.
+
+#### 15.14.5.3 Retrieve the current value
+
+An enumerator object’s `Current` property is affected by `yield return` statements in the iterator block.
+
+ > *Note*: The `Current` property is a synchronous property for both synchronous and asynchronous iterator objects. *end note*
+
+When an enumerator object is in the **suspended** state, the value of `Current` is the value set by the previous call to `MoveNext`. When an enumerator object is in the **before**, **running**, or **after** states, the result of accessing `Current` is unspecified.
+
+For an iterator with a yield type other than `object`, the result of accessing `Current` through the enumerator object’s `IEnumerable` implementation corresponds to accessing `Current` through the enumerator object’s `IEnumerator<T>` implementation and casting the result to `object`.
+
+#### 15.14.5.4 Dispose of resources
+
+The `Dispose` or `DisposeAsync` method is used to clean up the iteration by bringing the enumerator object to the **after** state.
+
+- If the state of the enumerator object is **before**, invoking `Dispose` changes the state to **after**.
+- If the state of the enumerator object is **running**, the result of invoking `Dispose` is unspecified.
+- If the state of the enumerator object is **suspended**, invoking `Dispose`:
+  - Changes the state to **running**.
+  - Executes any finally blocks as if the last executed `yield return` statement were a `yield break` statement. If this causes an exception to be thrown and propagated out of the iterator body, the state of the enumerator object is set to **after** and the exception is propagated to the caller of the `Dispose` method.
+  - Changes the state to **after**.
+- If the state of the enumerator object is **after**, invoking `Dispose` has no affect.
+
+### 15.14.6 Enumerable objects
+
+#### 15.14.6.1 General
+
+When a function member or local function returning an enumerable interface type is implemented using an iterator block, invoking the function member does not immediately execute the code in the iterator block. Instead, an ***enumerable object*** is created and returned.
+
+The enumerable object’s `GetEnumerator` or `GetAsyncEnumerator` method returns an enumerator object that encapsulates the code specified in the iterator block, and execution of the code in the iterator block occurs when the enumerator object’s `MoveNext` or `MoveNextAsync` method is invoked. An enumerable object has the following characteristics:
+
+- It implements `IEnumerable` and `IEnumerable<T>` or `IAsyncEnumerable<T>`, where `T` is the yield type of the iterator.
```

Comment:
We should probably clarify that the bracketing here is (`IEnumerable` and `IEnumerable<T>`) or (`IAsyncEnumerable<T>`). I don't have a suggestion for that right now.